lib/dal/src/func/binding/attribute.rs

```diff
@@ -98,37 +126,87 @@ impl AttributeBinding {
                 Some(EventualParent::SchemaVariant(schema_variant.into()))
             }
             (Some(component_id), None) => Some(EventualParent::Component(component_id.into())),
-            (Some(component_id), Some(schema_variant)) => {
+            (Some(component_id), Some(schema_variant_id)) => {
                 if Component::schema_variant_id(ctx, component_id.into()).await?
-                    == schema_variant.into()
+                    == schema_variant_id.into()
                 {
-                    Some(EventualParent::SchemaVariant(schema_variant.into()))
+                    Some(EventualParent::SchemaVariant(schema_variant_id.into()))
                 } else {
                     return Err(FuncBindingError::MalformedInput(
-                        "component and schema variant mismatch".to_owned(),
+                        AttributeBindingMalformedInput::EventualParentComponentNotFromSchemaVariant(
+                            component_id,
+                            schema_variant_id,
+                        ),
                     ));
                 }
             }
         };
         Ok(eventual_parent)
     }
-    pub fn assemble_attribute_output_location(
+
+    pub fn assemble_attribute_input_location(
         prop_id: Option<si_events::PropId>,
-        output_socket_id: Option<si_events::OutputSocketId>,
-    ) -> FuncBindingResult<AttributeFuncDestination> {
-        let output_location = match (prop_id, output_socket_id) {
-            (None, Some(output_socket_id)) => {
-                AttributeFuncDestination::OutputSocket(output_socket_id.into())
+        input_socket_id: Option<si_events::InputSocketId>,
+        static_argument_value: Option<serde_json::Value>,
+    ) -> FuncBindingResult<AttributeFuncArgumentSource> {
+        match (prop_id, input_socket_id, static_argument_value) {
+            (Some(prop_id), None, None) => Ok(AttributeFuncArgumentSource::Prop(prop_id.into())),
+            (None, Some(input_socket_id), None) => Ok(AttributeFuncArgumentSource::InputSocket(
+                input_socket_id.into(),
+            )),
+            (None, None, Some(static_argument_value)) => Ok(
+                AttributeFuncArgumentSource::StaticArgument(static_argument_value),
+            ),
+            (Some(prop_id), Some(input_socket_id), Some(static_argument_value)) => {
+                Err(FuncBindingError::MalformedInput(
+                    AttributeBindingMalformedInput::InputLocationAllOptionsProvided(
+                        prop_id,
+                        input_socket_id,
+                        static_argument_value,
+                    ),
+                ))
             }
-
-            (Some(prop_id), None) => AttributeFuncDestination::Prop(prop_id.into()),
-            _ => {
-                return Err(FuncBindingError::MalformedInput(
-                    "cannot set more than one output location".to_owned(),
```

Comment:
Similar issue here. Yes, the "output" location string is correct, but it also bundles the "no option provided" case with the "multiple options provided" case.